### PR TITLE
feat: add artgc podspecpatch fixes #11485

### LINKED ISSUE #11485: resources for the artgc pod
# Summary

Ability to set resources for the artifact garbage collection (artgc) pod.

Currently, the resources are [hard coded](https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/artifact_gc.go#L450) (to 64Mi mem and 0.1 cpu).

I could imagine being able to provide a `podSpecPatch` for the artifactgc pod with each workflow-template. And falling back to the defaults in [`mainContainer.resources`](https://argoproj.github.io/argo-workflows/workflow-controller-configmap.yaml). But perhaps there are other ways that make more sense.

@juliev0 , do you have any ideas/opinion about this as the author of the artifact gc code?

# Use Cases

We have workflows with many artifacts (magnitude of millions). The artifact garbage collection pods (`*-artgc-*`) for these get OOMKilled. Leaving the workflow in an undeletable (unless forced) state and leaving behind uncollected garbage artifacts.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by roelarents**
Fixes #11485

### Motivation

To be able to set different resources for the artgc pod.

### Modifications

Added `WorkflowLevelArtifactGC.PodSpecPatch` as discussed in #11485

Note: the patch could patch anything in the pod spec, including maybe undesired things. Seems the responsibility of the user to me.
Note 2: parameters are not substituted in the (patched) pod spec, like with the podSpecPatch for workflows.

### Verification

Expanded the `controller.TestProcessArtifactGCStrategy` unit test

### CODE REVIEW

**Review by juliev0 — COMMENTED** [review]
So far this looks good to me. 
Nit: can we rename `podAccessInfo` everywhere in this file to something else since it no longer pertains just to access? (something like `uniquePodInfo` or whatever you think)

**Comment by roelarents** [comment]
Sure. I made it just `podInfo`.

File: workflow/util/util.go [hunk 1]
```diff
@@ -1232,6 +1232,36 @@ func PodSpecPatchMerge(wf *wfv1.Workflow, tmpl *wfv1.Template) (string, error) {
...
+	}
+
+	var newPodSpec apiv1.PodSpec
+	err = json.Unmarshal(modJson, &newPodSpec)
+	if err != nil {
+		return nil, errors.Wrap(err, "", "Error in Unmarshalling after merge the patch")
+	}
+	return &newPodSpec, nil
+}
+
```
[the remark this answers is not in the document]
Comment (juliev0) [reply workflow/util/util.go:1264]: oh, and thanks for adding the comments that weren't there in the original code, too

**Review by juliev0 — APPROVED** [review]

**Review by terrytangyuan — COMMENTED** [review]
Could you run `make codegen`?

**Review by juliev0 — DISMISSED** [review]
I approved but I guess you need to `make codegen` first

**Comment by roelarents** [comment]
Thx. I'm trying to do the code generation. But I'm running into `k8s.io/apimachinery/pkg/runtime/schema/generated.proto: File not found.`. Similar as [here](https://github.com/argoproj/argo-workflows/pull/9235#issuecomment-1198125583). So I might also ask for you to do it if I cannot make it work 😔 🙏 .

**Comment by roelarents** [comment]
> Sure. I think you'll need to give me permissions to write to your fork?

Thx and done. I've invited you as collaborator. Also _Allow edits and access to secrets by maintainers_ is on.

**Comment by juliev0** [comment]
Well, I fixed the codegen but I'm not sure what's going on with the CI now. Many of the jobs seem to be stuck in "Waiting to run this check...". Any thoughts @terrytangyuan ?

**Comment by juliev0** [comment]
> Well, I fixed the codegen but I'm not sure what's going on with the CI now. Many of the jobs seem to be stuck in "Waiting to run this check...". Any thoughts @terrytangyuan ?

Looks like it finally ran the tests - not sure why that took so long...

**Comment by terrytangyuan** [comment]
Triggered re-run.

**Comment by juliev0** [comment]
@roelarents Sorry about all the flakey e2e tests. Can you try doing some empty commits and see if you can get it to pass?

**Review by terrytangyuan — APPROVED** [review]

**Pull request merged**